### PR TITLE
ceph: add support in RGW to communicate vault with TLS

### PR DESCRIPTION
**Pull request by thotz**
**Description of your changes:**

From ceph v16.2.6 onwards the vault TLS suppport in RGW was added,
include similar changes for RGW.

Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: tests/scripts/deploy-validate-vault.sh [hunk 1]
```diff
@@ -132,7 +132,10 @@ function validate_rgw_token {
   RGW_TOKEN_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-token-file" | cut -f2- -d=)
   VAULT_PATH_PREFIX=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-prefix" | cut -f2- -d=)
   VAULT_TOKEN=$(kubectl -n rook-ceph exec $RGW_POD -- cat $RGW_TOKEN_FILE)
-  
+  #VAULT_CACERT=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-ssl-cacert" | cut -f2- -d=)
```
Comment (travisn) [inline tests/scripts/deploy-validate-vault.sh:135]: Are these test commands still WIP?

**Review by travisn — CHANGES_REQUESTED** [review]

File: tests/scripts/deploy-validate-vault.sh (hunk 1 above)
Comment (thotz) [reply tests/scripts/deploy-validate-vault.sh:135]: @travisn The required changes will be available on ceph v16.2.6, which is not yet released. Or else we need to add check in the test script to see whether it is ceph version is higher or not

Comment (leseb) [reply tests/scripts/deploy-validate-vault.sh:135]: @thotz can you try with `ceph/daemon-base:latest-pacific-devel` to validate your patch?

**Review by travisn — APPROVED** [review]

File: pkg/daemon/ceph/osd/kms/volumes.go [hunk 2]
```diff
@@ -82,25 +82,32 @@ func VaultVolumeAndMount(config map[string]string) (v1.Volume, v1.VolumeMount) {
...
 	return ""
 }
 
 // VaultTokenFileVolume save token from secret as volume mount
-func VaultTokenFileVolume(tokenSecretName string) v1.Volume {
-	return v1.Volume{
-		Name: secrets.TypeVault,
+func VaultTokenFileVolume(tokenSecretName string) (v1.Volume, v1.VolumeMount) {
+	volume := v1.Volume{
+		Name: "vault-file-vol",
```
Comment (leseb) [inline pkg/daemon/ceph/osd/kms/volumes.go:99]: Why not keeping `secrets.TypeVault`?

File: .github/workflows/canary-integration-test.yml [hunk 3]
```diff
@@ -576,9 +576,11 @@ jobs:
         yq merge --inplace --arrays append tests/manifests/test-cluster-on-pvc-encrypted.yaml tests/manifests/test-kms-vault-spec.yaml
         yq write -i tests/manifests/test-cluster-on-pvc-encrypted.yaml "spec.storage.storageClassDeviceSets[0].count" 2
         yq write -i tests/manifests/test-cluster-on-pvc-encrypted.yaml "spec.storage.storageClassDeviceSets[0].volumeClaimTemplates[0].spec.resources.requests.storage" 6Gi
+        yq write -i tests/manifests/test-cluster-on-pvc-encrypted.yaml "spec.cephVersion.image" quay.io/ceph/daemon-base:latest-pacific-devel
         kubectl create -f tests/manifests/test-cluster-on-pvc-encrypted.yaml
         yq merge --inplace --arrays append tests/manifests/test-object.yaml tests/manifests/test-kms-vault-spec.yaml
         sed -i 's/ver1/ver2/g' tests/manifests/test-object.yaml
+        sed -i 's/true/false/g' tests/manifests/test-object.yaml
```
[the remark this answers is not in the document]
Comment (thotz) [reply .github/workflows/canary-integration-test.yml:583]: I am changing `VAULT_SKIP_VERIFY` into `false`, not `preservePoolsOnDelete`. Is okay to remove it? When I removed it failed for OSD creation AFAIR.  So I made this change, so it won;t effect the existing case

File: pkg/daemon/ceph/osd/kms/volumes.go (hunk 2 above)
Comment (thotz) [reply pkg/daemon/ceph/osd/kms/volumes.go:99]: @leseb : If TLS enabled then it will end up having two volumes having the same name, one with TLS certs for vault and the other with the token file. I didn't modify `VaultVolumeAndMount()` to include the `token` file since it is not required for the OSD encryption and wrote separate function for that

File: .github/workflows/canary-integration-test.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (leseb) [reply .github/workflows/canary-integration-test.yml:583]: No it's just that the `sed` is not really clear since you are changing all the occurrences of `true` with `false`. Can you just isolate that so we know what the CI does?

**Review by leseb — APPROVED** [review]

**Comment by thotz** [comment]
@leseb: anyidea what is failing here, locally when I tried steps in `github actions job` it is working fine. Is OSD not coming up os the issue or toolpods is missing I am not able to figure this out

**Comment by leseb** [comment]
> @leseb: anyidea what is failing here, locally when I tried steps in `github actions job` it is working fine. Is OSD not coming up os the issue or toolpods is missing I am not able to figure this out

```
2021-08-31 12:51:41.509712 E | ceph-cluster-controller: failed to reconcile CephCluster "rook-ceph/rook-ceph". failed to reconcile cluster "rook-ceph": failed to configure local ceph cluster: failed to perform validation before cluster creation: failed to validate kms connection details: failed to get backend version: failed to initialize vault client: open vault-client-cert: no such file or directory
```

**Review by leseb — CHANGES_REQUESTED** [review]
holding until CI job is fixed

**Comment by thotz** [comment]
> > @leseb: anyidea what is failing here, locally when I tried steps in `github actions job` it is working fine. Is OSD not coming up os the issue or toolpods is missing I am not able to figure this out
> 
> ```
> 2021-08-31 12:51:41.509712 E | ceph-cluster-controller: failed to reconcile CephCluster "rook-ceph/rook-ceph". failed to reconcile cluster "rook-ceph": failed to configure local ceph cluster: failed to perform validation before cluster creation: failed to validate kms connection details: failed to get backend version: failed to initialize vault client: open vault-client-cert: no such file or directory
> ```

@leseb : from code reading even though `tlsConfig.Insecure` is set ConfigureTLS() , setting `VAULT_TLS_SECRET` may creating the issue here. If understand correctly `tlsConfig.other` options expects path we but are providing secret names. IMO NewClient() is not ignoring this values  even though `tlsConfig.Insecure` is set

File: pkg/operator/ceph/object/spec.go [hunk 4]
```diff
@@ -309,6 +310,25 @@ func (c *clusterConfig) makeDaemonContainer(rgwConfig *rgwConfig) v1.Container {
 					c.store.Spec.Security.KeyManagementService.ConnectionDetails[kms.VaultSecretEngineKey]),
 			)
 		}
+		if c.store.Spec.Security.KeyManagementService.IsTLSEnabled() &&
+			c.clusterInfo.CephVersion.IsAtLeast(cephver.CephVersion{Major: 16, Minor: 2, Extra: 5}) {
+			/*			_, vaultTLSVolMount := kms.VaultVolumeAndMount(c.store.Spec.Security.KeyManagementService.ConnectionDetails)
```
Comment (travisn) [inline pkg/operator/ceph/object/spec.go:315]: Should this commented be deleted, or still investigating if it's needed?

Comment (thotz) [reply pkg/operator/ceph/object/spec.go:315]: @travisn mybad, forgot to remove

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
No unit tests?

**Comment by thotz** [comment]
> No unit tests?

I can add unit test for VaultTokenFileVolume() similar to TestTLSSecretVolumeAndMount().

Other addition can done in TestCheckRGWKMS() to validate the tls client certs passed, but similar cases are performed in TestValidateConnectionDetails() and CheckRGWKMS() calls the ValidateConnectionDetails() internally. So do I need to cover these cases?

**Comment by BlaineEXE** [comment]
> > No unit tests?
> > @BlaineEXE  I can add unit test for VaultTokenFileVolume() similar to TestTLSSecretVolumeAndMount().
> 
> Other addition can done in TestCheckRGWKMS() to validate the tls client certs passed, but similar cases are performed in TestValidateConnectionDetails() and CheckRGWKMS() calls the ValidateConnectionDetails() internally. So do I need to cover these cases?

Yes. All code that adds or modifies a feature should come with added or modified unit tests. Part of development is writing tests to ensure the feature works as intended and to catch corner case issues during development. We have collectively (including myself) slacked off on doing this for Rook in the past, but with as much complexity as Rook has and the importance that it works well for our downstream customers and upstream users, we can't continue to ignore that responsibility.

When I write a new feature, the unit tests I write for it **always** catch bugs that I make in the code. Every single time. I usually find that I spend more time working on tests than the feature itself and that writing good tests that cover all corner cases is much more difficult than implementing the feature. When I estimate the time it will take to implement a feature, I often estimate the time I think it will take to implement the feature itself, then multiply by three to estimate the time it will take me to implement the feature plus the tests (I assume tests take twice as long).

Questions that every unit test should answer and test (not all are always applicable):
1. What if the feature is enabled?
2. What if the feature is disabled?
3. What if the feature is only partially enabled? (and how many ways can it be partially enabled?)
4. What errors can be encountered when the configuration is being applied? (need test for each kind of error)
5. Is there a slice/array involved? (test slice length=0, length=1, length=3, length=max, larger than max length)
6. What if an input is not specified? (test for every input that can be entered incorrectly)
7. What if an input is specified incorrectly? (test every input that can be incorrect)
8. What if a resource we rely on doesn't exist?

I'm sure there are some questions I missed.

**Review by travisn — COMMENTED** [review]
It's great to see all the unit and integration tests, thanks

File: pkg/operator/ceph/object/spec.go [hunk 5]
```diff
@@ -309,6 +310,23 @@ func (c *clusterConfig) makeDaemonContainer(rgwConfig *rgwConfig) v1.Container {
...
+			}
+			if kms.GetParam(c.store.Spec.Security.KeyManagementService.ConnectionDetails, api.EnvVaultClientKey) != "" {
+				container.Args = append(container.Args,
+					cephconfig.NewFlag("rgw crypt vault ssl clientkey", path.Join(c.DataPathMap.ContainerDataDir, kms.VaultKeyFileName)))
+			}
+			if kms.GetParam(c.store.Spec.Security.KeyManagementService.ConnectionDetails, api.EnvVaultCACert) != "" {
+				container.Args = append(container.Args,
+					cephconfig.NewFlag("rgw crypt vault ssl cacert", path.Join(c.DataPathMap.ContainerDataDir, kms.VaultCAFileName)))
+			}
+		}
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/spec.go:329]: Anything placed in the ContainerDataDir could be persisted to the host, and it is for information only for the specific Ceph daemon. Similarly to how Vault is configured for OSDs, there should be a completely separate volume mount for these keys and certs.

You should look to see if you can reuse any code found in `pkg/daemon/ceph/osd/kms/volumes.go`. If so, perhaps we should have move reusable bits to `pkg/daemon/ceph/kms` or someplace similar. What do you think @leseb ?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/kms/volumes_test.go [hunk 6]
```diff
@@ -77,3 +77,43 @@ func TestTLSSecretVolumeAndMount(t *testing.T) {
...
+	tests := []struct {
+		name string
+		args args
+		want []v1.VolumeProjection
+	}{
+		{"empty", args{tokenSecretName: "", config: map[string]string{"foo": "bar"}}, []v1.VolumeProjection{}},
+		{"token file", args{tokenSecretName: "vault-token", config: map[string]string{"foo": "bar"}}, []v1.VolumeProjection{
+			{Secret: &v1.SecretProjection{LocalObjectReference: v1.LocalObjectReference{Name: "vault-token"}, Items: []v1.KeyToPath{{Key: "token", Path: "vault.token", Mode: &m}}, Optional: nil}}},
+		},
+		{"token and single ca", args{tokenSecretName: "vault-token", config: map[string]string{"VAULT_CACERT": "vault-ca-secret"}}, []v1.VolumeProjection{
```
[the remark this answers is not in the document]
Comment (thotz) [reply pkg/daemon/ceph/osd/kms/volumes_test.go:95]: The message should be more like only ca(not with cert and key)

File: pkg/operator/ceph/object/spec.go (hunk 5 above)
Comment (thotz) [reply pkg/operator/ceph/object/spec.go:329]: Sorry I was not aware of that. Another option is to add these in "/etc/ceph" code path, but all the files owned in that location was by `root`. Here the file needs to be owned by `ceph:ceph` so that the RGW process can access it and it is done via init container

Comment (BlaineEXE) [reply pkg/operator/ceph/object/spec.go:329]: It looks like Seb configured the KMS for OSDs to use `/etc/vault`

Comment (thotz) [reply pkg/operator/ceph/object/spec.go:329]: @BlaineEXE : By default, secrets are mounted read-only files with `root:root`and RGW needs this to be `ceph` owned file. I cannot chown on this mounted file or add FSGroup to `ceph` podspec(in openshift this value will be set by default). So I am copying these files to `containerdir` and then perform chown via init container(refer `setupVaultTokenFile`). By default file is mounted in the same location `/etc/vault`.

Comment (BlaineEXE) [reply pkg/operator/ceph/object/spec.go:329]: I think this is probably also a requirement of KMS on OSDs, but @leseb would know for sure. Having KMS handling/configuration that is shared between rook-ceph components will make things easier to maintain, debug, and unit test.

File: pkg/daemon/ceph/osd/kms/volumes_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply pkg/daemon/ceph/osd/kms/volumes_test.go:95]:
```suggestion
		{"token and ca", args{tokenSecretName: "vault-token", config: map[string]string{"VAULT_CACERT": "vault-ca-secret"}}, []v1.VolumeProjection{
```

File: pkg/operator/ceph/object/spec.go (hunk 5 above)
Comment (thotz) [reply pkg/operator/ceph/object/spec.go:329]: @BlaineEXE : I cannot create this new directory in `/etc/ceph` or `/etc/vault` because they are already mounted in the rgw pod as read-only secret. If I try to create the directory via init container, it won't be persisted for the main rgw container, so approach to save it  `ContainerDataDir`. Another way to resolve this issue is in ceph, remove restrictions on the token file for the vault(currently it restricts to have permission to only to the owner of the file, not even `read` permission to group/others) but not sure whether it will acceptable from the ceph community

File: Documentation/ceph-object-store-crd.md [hunk 7]
```diff
@@ -197,8 +197,6 @@ For RGW, please note the following:
 $ vault kv put rook/<mybucketkey> key=$(openssl rand -base64 32) # kv engine
 $ vault write -f transit/keys/<mybucketkey> exportable=true # transit engine
 
-* TLS authentication with custom certs between Vault and RGW are yet to be supported.
```
Comment (leseb) [inline Documentation/ceph-object-store-crd.md:200]: are supported as of "version XXX"?

File: pkg/operator/ceph/object/spec.go [hunk 8]
```diff
@@ -45,11 +46,11 @@ set -e
 VAULT_TOKEN_OLD_PATH=%s
 VAULT_TOKEN_NEW_PATH=%s
 
-cp --verbose $VAULT_TOKEN_OLD_PATH $VAULT_TOKEN_NEW_PATH
+cp -r --verbose $VAULT_TOKEN_OLD_PATH/..data/. $VAULT_TOKEN_NEW_PATH
```
Comment (leseb) [inline pkg/operator/ceph/object/spec.go:49]:
```suggestion
cp --recursive --verbose $VAULT_TOKEN_OLD_PATH/..data/. $VAULT_TOKEN_NEW_PATH
```

File: pkg/operator/ceph/object/spec.go [hunk 9]
```diff
@@ -45,11 +46,11 @@ set -e
...
 
-chmod --verbose 400 $VAULT_TOKEN_NEW_PATH
-
-chown --verbose ceph:ceph $VAULT_TOKEN_NEW_PATH
+chmod -R --verbose 400 $VAULT_TOKEN_NEW_PATH/*
```
Comment (leseb) [inline pkg/operator/ceph/object/spec.go:51]:
```suggestion
chmod --recursive --verbose 400 $VAULT_TOKEN_NEW_PATH/*
```

File: pkg/operator/ceph/object/spec.go [hunk 10]
```diff
@@ -45,11 +46,11 @@ set -e
...
+chmod  --verbose 700 $VAULT_TOKEN_NEW_PATH
+chown -R --verbose ceph:ceph $VAULT_TOKEN_NEW_PATH
```
Comment (leseb) [inline pkg/operator/ceph/object/spec.go:53]:
```suggestion
chown --recursive --verbose ceph:ceph $VAULT_TOKEN_NEW_PATH
```

File: .github/workflows/canary-integration-test.yml [hunk 11]
```diff
@@ -576,9 +576,8 @@ jobs:
         yq merge --inplace --arrays append tests/manifests/test-cluster-on-pvc-encrypted.yaml tests/manifests/test-kms-vault-spec.yaml
         yq write -i tests/manifests/test-cluster-on-pvc-encrypted.yaml "spec.storage.storageClassDeviceSets[0].count" 2
         yq write -i tests/manifests/test-cluster-on-pvc-encrypted.yaml "spec.storage.storageClassDeviceSets[0].volumeClaimTemplates[0].spec.resources.requests.storage" 6Gi
+        yq write -i tests/manifests/test-cluster-on-pvc-encrypted.yaml "spec.cephVersion.image" quay.io/ceph/daemon-base:latest-pacific-devel
```
Comment (leseb) [inline .github/workflows/canary-integration-test.yml:579]: 16.2.6 is out and will soon be consumed in https://github.com/rook/rook/pull/8743 so let's wait for it, still good to validate testing though

File: tests/manifests/test-object.yaml [hunk 12]
```diff
@@ -18,5 +18,17 @@ spec:
   preservePoolsOnDelete: false
   gateway:
     port: 80
-    # securePort: 443
     instances: 1
+  security:
```
Comment (leseb) [inline tests/manifests/test-object.yaml:22]: Let's have this section in a dedicated file like `tests/manifests/test-kms-vault-spec.yaml` and append it to `tests/manifests/test-object.yaml` during the integration test.

File: tests/manifests/test-object.yaml [hunk 13]
```diff
@@ -18,5 +18,17 @@ spec:
...
+    kms:
+      connectionDetails:
+        KMS_PROVIDER: vault
+        VAULT_ADDR: https://vault.default.svc.cluster.local:8200
+        VAULT_BACKEND_PATH: rook/ver2
+        VAULT_BACKEND: "v2"
+        VAULT_SECRET_ENGINE: kv
+        VAULT_SKIP_VERIFY: "false"
```
Comment (leseb) [inline tests/manifests/test-object.yaml:30]: How did you get the signed certificate? Should this be true?

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/object/spec.go (hunk 5 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/spec.go]: This content **cannot** go into `ContainerDataDir` for the reasons I explained. It is wholly untrue that this is the only feasible approach. Examine the OSD KMS code that Seb wrote (that this PR should reuse as much as possible), and see that `/etc/vault` is mounted from an `emptyDir`. There is no need to make changes to Ceph.

File: tests/manifests/test-object.yaml (hunk 13 above)
Comment (thotz) [reply tests/manifests/test-object.yaml:30]: Sorry I didn't understand your question, this certificate was validated by k8s right. So no need of skipping the vault check(and secondly Rook <-> vault communication is not required for the RGW. if `kvengine` version checks fail RGW assumes engine is v2 in the `CheckRGWKMS()`

File: tests/manifests/test-object.yaml (hunk 12 above)
Comment (thotz) [reply tests/manifests/test-object.yaml:22]: If I use dedicated session from `tests/manifests/test-kms-vault-spec.yaml`, then I end up adding/modifying following fields for `test-object.yaml` by calling sed/yq
```  
      VAULT_BACKEND: "v2"
        VAULT_SKIP_VERIFY: "false"
        VAULT_CACERT: vault-ca-cert
        VAULT_CLIENT_CERT: vault-client-cert
        VAULT_CLIENT_KEY: vault-client-key
```
To avoid this I have added entire block again in `test-object.yaml`

File: pkg/daemon/ceph/osd/kms/volumes.go [hunk 14]
```diff
@@ -82,25 +82,36 @@ func VaultVolumeAndMount(config map[string]string) (v1.Volume, v1.VolumeMount) {
...
 
 	}
 
 	return ""
 }
 
 // VaultTokenFileVolume save token from secret as volume mount
-func VaultTokenFileVolume(tokenSecretName string) v1.Volume {
-	return v1.Volume{
+func VaultTokenFileVolume(tokenSecretName string, config map[string]string) v1.Volume {
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/kms/volumes.go:97]: What config is being referenced here? I don't understand from the function signature or comment what config should be used as input here.

File: pkg/daemon/ceph/osd/kms/volumes.go [hunk 15]
```diff
@@ -82,25 +82,36 @@ func VaultVolumeAndMount(config map[string]string) (v1.Volume, v1.VolumeMount) {
...
-					{Key: KMSTokenSecretNameKey, Path: VaultFileName},
-				}}}}
+			Projected: &v1.ProjectedVolumeSource{
+				Sources: secretVolumeProjections,
+			},
+		},
+	}
+
+	return volume
 }
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/kms/volumes.go:117]: In the interest of code sharing, can `VaultVolumeAndMount` be reused here and `VaultTokenFileVolume` be removed? Thoughts @leseb ?

File: pkg/daemon/ceph/osd/kms/volumes_test.go [hunk 16]
```diff
@@ -77,3 +77,43 @@ func TestTLSSecretVolumeAndMount(t *testing.T) {
 		})
 	}
 }
+func TestVaultTokenVolumeAndMount(t *testing.T) {
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/kms/volumes_test.go:80]: Tests should be named `TestFunction` where "Function" is the name of the function under test.

```suggestion
func TestVaultTokenVolume(t *testing.T) {
```

File: pkg/operator/ceph/object/spec_test.go [hunk 17]
```diff
@@ -379,4 +379,34 @@ func TestCheckRGWKMS(t *testing.T) {
...
+		assert.EqualError(t, err, "failed to validate vault connection details: failed to find TLS connection key \"cert\" for \"VAULT_CACERT\" in k8s secret \"vault-ca-secret\"")
+	})
+	t.Run("TLS config is valid", func(t *testing.T) {
+		tlsSecret.Data = map[string][]byte{"cert": []byte("envnrevbnbvsbjkrtn")}
+		_, err = context.Clientset.CoreV1().Secrets(store.Namespace).Update(ctx, tlsSecret, metav1.UpdateOptions{})
+		assert.NoError(t, err)
+		b, err = c.CheckRGWKMS()
+		assert.True(t, b)
+		assert.NoError(t, err, "")
+	})
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/spec_test.go:411]: All of these tests share the same environment, which we should avoid where possible. It makes tests harder to read, and it can make them fragile when code is updated in the future. Please refactor this test to make each test run independent. This goes for the previously-existing tests as well. 

It should be easy to create a test setup method to reuse for each test

```go
func TestCheckRGWKMS(t *testing.T) {
    setupTest := func() *cluster.Config {
        /* do test setup and return whatever you need for tests */
    }
    
    t.Run("test thing", func(t *testing.T) {
        clusterConfig := setupTest()
        
        // make any special modifications to the config for this test....
        c.store.Spec.Security.KeyManagementService.ConnectionDetails = stuff
        clusterConfig.context.Clientset.CoreV1().CreateSomeResource()
        
        ret, err := clusterConfig.CheckRGWKMS()
        assert.Equal(t, "something", ret)
    }
    
    /* etc. */
}

```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
We should also run a version of `TestPodSpecs` for the RGW spec where KMS is enabled.

File: .github/workflows/canary-integration-test.yml [hunk 18]
```diff
@@ -422,8 +422,8 @@ jobs:
         yq write -i tests/manifests/test-cluster-on-pvc-encrypted.yaml "spec.storage.storageClassDeviceSets[0].volumeClaimTemplates[0].spec.resources.requests.storage" 6Gi
         kubectl create -f tests/manifests/test-cluster-on-pvc-encrypted.yaml
         yq merge --inplace --arrays append tests/manifests/test-object.yaml tests/manifests/test-kms-vault-spec-token-auth.yaml
-        sed -i 's/ver1/ver2/g' tests/manifests/test-object.yaml
-        kubectl create -f tests/manifests/test-object.yaml
```
Comment (leseb) [inline .github/workflows/canary-integration-test.yml:426]: Why is this line gone?

File: .github/workflows/canary-integration-test.yml [hunk 19]
```diff
@@ -422,8 +422,8 @@ jobs:
...
+        yq write -i tests/manifests/test-object.yaml "spec.security.kms.connectionDetails.VAULT_BACKEND_PATH" rook/ver2
+        yq write -i tests/manifests/test-object.yaml "spec.security.kms.connectionDetails.VAULT_BACKEND" v2
```
Comment (leseb) [inline .github/workflows/canary-integration-test.yml:426]: Let's not pass `VAULT_BACKEND` since it's auto-detected.

File: pkg/operator/ceph/object/spec.go [hunk 20]
```diff
@@ -45,11 +46,11 @@ set -e
...
 VAULT_TOKEN_NEW_PATH=%s
 
-cp --verbose $VAULT_TOKEN_OLD_PATH $VAULT_TOKEN_NEW_PATH
+cp --recursive --verbose $VAULT_TOKEN_OLD_PATH/..data/. $VAULT_TOKEN_NEW_PATH
 
-chmod --verbose 400 $VAULT_TOKEN_NEW_PATH
-
-chown --verbose ceph:ceph $VAULT_TOKEN_NEW_PATH
+chmod --recursive --verbose 400 $VAULT_TOKEN_NEW_PATH/*
+chmod  --verbose 700 $VAULT_TOKEN_NEW_PATH
```
Comment (leseb) [inline pkg/operator/ceph/object/spec.go:52]:
```suggestion
chmod --verbose 700 $VAULT_TOKEN_NEW_PATH
```

File: tests/scripts/deploy-validate-vault.sh [hunk 21]
```diff
@@ -157,9 +157,13 @@ function validate_rgw_token {
...
   VAULT_PATH_PREFIX=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-prefix" | cut -f2- -d=)
   VAULT_TOKEN=$(kubectl -n rook-ceph exec $RGW_POD -- cat $RGW_TOKEN_FILE)
+  VAULT_CACERT_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-ssl-cacert" | cut -f2- -d=)
+  VAULT_CLIENT_CERT_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-ssl-clientcert" | cut -f2- -d=)
+  VAULT_CLIENT_KEY_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-ssl-clientkey" | cut -f2- -d=)
+
 
   #fetch key from vault server using token from RGW pod, P.S using -k for curl since custom ssl certs not yet to support in RGW
-  FETCHED_KEY=$(kubectl -n rook-ceph exec $RGW_POD -- curl -k -X GET -H "X-Vault-Token:$VAULT_TOKEN" "$VAULT_SERVER""$VAULT_PATH_PREFIX"/"$RGW_BUCKET_KEY"|jq -r .data.data.key)
+  FETCHED_KEY=$(kubectl -n rook-ceph exec $RGW_POD -- curl  --key "$VAULT_CLIENT_KEY_FILE" --cert "$VAULT_CLIENT_CERT_FILE" --cacert "$VAULT_CACERT_FILE" -X GET -H "X-Vault-Token:$VAULT_TOKEN" "$VAULT_SERVER""$VAULT_PATH_PREFIX"/"$RGW_BUCKET_KEY"|jq -r .data.data.key)
```
Comment (leseb) [inline tests/scripts/deploy-validate-vault.sh:166]:
```suggestion
  FETCHED_KEY=$(kubectl -n rook-ceph exec $RGW_POD -- curl --key "$VAULT_CLIENT_KEY_FILE" --cert "$VAULT_CLIENT_CERT_FILE" --cacert "$VAULT_CACERT_FILE" -X GET -H "X-Vault-Token:$VAULT_TOKEN" "$VAULT_SERVER""$VAULT_PATH_PREFIX"/"$RGW_BUCKET_KEY"|jq -r .data.data.key)
```

File: pkg/daemon/ceph/osd/kms/volumes.go [hunk 22]
```diff
@@ -29,16 +29,16 @@ const (
...
+	VaultKeyFileName  = "vault.key"
 
 	// File name for token file
 	VaultFileName = "vault.token"
 )
 
-// TLSSecretVolumeAndMount return the volume and matching volume mount for mounting the secrets into /etc/vault
-func TLSSecretVolumeAndMount(config map[string]string) []v1.VolumeProjection {
+// VaultSecretVolumeAndMount return the volume and matching volume mount for mounting the vault secrets into /etc/vault
+func VaultSecretVolumeAndMount(kmsVaultConfigFiles map[string]string, tokenSecretName string, mountToken bool) []v1.VolumeProjection {
```
Comment (leseb) [inline pkg/daemon/ceph/osd/kms/volumes.go:41]: Why do we need a bool here? Can't we just check `tokenSecretName != ""`?

File: pkg/daemon/ceph/osd/kms/volumes.go [hunk 23]
```diff
@@ -49,25 +49,30 @@ func TLSSecretVolumeAndMount(config map[string]string) []v1.VolumeProjection {
...
+		projectionSecret.Name = tokenSecretName
+		secretProjection := v1.VolumeProjection{Secret: projectionSecret}
+		secretVolumeProjections = append(secretVolumeProjections, secretProjection)
+	}
 	return secretVolumeProjections
 }
 
 // VaultVolumeAndMount returns Vault volume and volume mount
-func VaultVolumeAndMount(config map[string]string) (v1.Volume, v1.VolumeMount) {
+func VaultVolumeAndMount(kmsVaultConfigFiles map[string]string, tokenSecretName string, mountToken bool) (v1.Volume, v1.VolumeMount) {
```
Comment (leseb) [inline pkg/daemon/ceph/osd/kms/volumes.go:70]: ditto

**Review by leseb — CHANGES_REQUESTED** [review]

**Review by BlaineEXE — COMMENTED** [review]
I flagged a few nit typos, given Seb already requested changes. I think this looks pretty good overall now, with some good questions from Seb.

File: tests/scripts/deploy-validate-vault.sh [hunk 24]
```diff
@@ -157,9 +157,13 @@ function validate_rgw_token {
   RGW_TOKEN_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-token-file" | cut -f2- -d=)
   VAULT_PATH_PREFIX=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-prefix" | cut -f2- -d=)
   VAULT_TOKEN=$(kubectl -n rook-ceph exec $RGW_POD -- cat $RGW_TOKEN_FILE)
+  VAULT_CACERT_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-ssl-cacert" | cut -f2- -d=)
```
Comment (leseb) [inline tests/scripts/deploy-validate-vault.sh:160]:
```suggestion
  VAULT_CACERT_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep "rgw-crypt-vault-ssl-cacert" | cut -f2- -d=)
```

File: tests/scripts/deploy-validate-vault.sh [hunk 25]
```diff
@@ -157,9 +157,13 @@ function validate_rgw_token {
...
+  VAULT_CLIENT_CERT_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-ssl-clientcert" | cut -f2- -d=)
```
Comment (leseb) [inline tests/scripts/deploy-validate-vault.sh:161]:
```suggestion
  VAULT_CLIENT_CERT_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep "rgw-crypt-vault-ssl-clientcert" | cut -f2- -d=)
```

File: tests/scripts/deploy-validate-vault.sh [hunk 26]
```diff
@@ -157,9 +157,13 @@ function validate_rgw_token {
...
+  VAULT_CLIENT_KEY_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep  "rgw-crypt-vault-ssl-clientkey" | cut -f2- -d=)
```
Comment (leseb) [inline tests/scripts/deploy-validate-vault.sh:162]:
```suggestion
  VAULT_CLIENT_KEY_FILE=$(kubectl -n rook-ceph describe pods "$RGW_POD" | grep "rgw-crypt-vault-ssl-clientkey" | cut -f2- -d=)
```

**Review by leseb — CHANGES_REQUESTED** [review]

**Review by leseb — APPROVED** [review]

**Review by BlaineEXE — APPROVED** [review]
